crates/matrix-sdk/src/authentication/oauth/qrcode/reciprocate/mod.rs

```diff
@@ -0,0 +1,222 @@
+    // -- MSC4108 OAuth 2.0 login step 4 continued
+    let request = device::get_device::v3::Request::new(device_id.to_base64().into());
+    if client.send(request).await.is_ok() {
+        channel
+            .send_json(QrAuthMessage::LoginFailure {
+                reason: LoginFailureReason::DeviceAlreadyExists,
+                homeserver: None,
+            })
+            .await?;
+        return Err(Error::DeviceIDAlreadyInUse);
+    }
+
+    // We emit an update so that the caller can open the verification URI in a
+    // system browser to consent to the login.
+    // -- MSC4108 OAuth 2.0 login step 4 continued
+    let verification_uri = Url::parse(
+        device_authorization_grant
+            .verification_uri_complete
+            .map(VerificationUriComplete::into_secret)
+            .unwrap_or(device_authorization_grant.verification_uri.to_string())
+            .as_str(),
+    )
+    .map_err(|_| Error::UnableToCreateDevice)?;
+    state.set(ReciprocateProgress::WaitingForAuth { verification_uri }.into());
+
+    // We send the new device the m.login.protocol_accepted message to let it know
+    // that the consent process is in progress.
+    // -- MSC4108 OAuth 2.0 login step 4 continued
+    let message = QrAuthMessage::LoginProtocolAccepted;
+    channel.send_json(&message).await?;
+
+    // The new device displays the user code it received from the authorization
+    // server and starts polling for an access token. In parallel, the user
+    // consents to the new login in the browser on this device, while verifying
+    // the user code displayed on the other device. -- MSC4108 OAuth 2.0 login
+    // steps 5 & 6
+
+    // We wait for the new device to send us the m.login.success message
+    let message: QrAuthMessage = channel.receive_json().await?;
+    let QrAuthMessage::LoginSuccess = message else {
+        return Err(Error::Unknown(
+            "Receiving unexpected message when expecting LoginSuccess".to_owned(),
+        ));
+    };
+
+    // We check that the new device was created successfully allowing for up to 10
+    // seconds of delay. -- MSC4108 Secret sharing and device verification step
+    // 1
+    let request = device::get_device::v3::Request::new(device_id.to_base64().into());
+    let deadline = Instant::now() + Duration::from_secs(10);
+    loop {
+        if client.send(request.clone()).await.is_ok() {
+            break;
+        }
+        // If the deadline hasn't yet passed, give it some time and retry the request.
+        if Instant::now() < deadline {
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            continue;
+        }
+        // The deadline has passed. Let's fail the login process.
+        channel
+            .send_json(QrAuthMessage::LoginFailure {
+                reason: LoginFailureReason::DeviceNotFound,
+                homeserver: None,
+            })
+            .await?;
+        return Err(Error::DeviceIDAlreadyInUse);
```

Comment:
This is an if/else if/else branch with the branches flattened out.

Can we make this a proper if/else if/else to make it clear from a glance that only one of the branches will be executed at each iteration.